File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& mat, int target) {
        //staircase method
        int r=0;
        int c=mat[0].size()-1;
        while(r<mat.size() && c>=0){
            if(mat[r][c] == target){
                return true;
            }
            else if(mat[r][c] <target){
                r++;
            }
            else{
                c--;
            }
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, mat: List[List[int]], target: int) -> bool:
        r = len(mat)
        c = len(mat[0])
        l=0
        h = r*c-1
        while(l<=h):
            mid =l+(h-l)//2
            row = mid //c
            col = mid % c
            if mat[row][col] == target:
                return True
            elif mat[row][col] < target:
                l = mid +1
            else:
                h = mid -1
        return False

       


        

        